docs/languages/csharp-iterators.md

```diff
@@ -0,0 +1,247 @@
+# Introduction
+
+Almost every program you write will have some need to iterate
+over a collection. You'll write code that examines every item in
+a collection. You'll also create classes
+that produce a collection by creating a method that produces an
+iterator for the elements of that class. The C# language provides
+features for both these scenarios. This article provides an overview
+of those features.
+
+# iterating with foreach
+
+Enumerating a collection is simple: The `foreach` keyword enumerates
+a collection, executing the embedded statement once for each element
+in the collection:
+ 
+```cs
+foreach (var item in collection)
+{
+   Console.WriteLine(item.ToString());
+}
+```
+
+That's all there is to it. To iterate over all the contents of a collection,
+the `foreach` statement is all you need. The `foreach` statement isn't magic,
+though. It relies on two interfaces defined in the .NET core library in order
+to generate the code necessary to iterate a collection: `IEnumerable<T>` and
+`IEnumerator<T>`. 
+
+Both of these interfaces also have non-generic counterparts: `IEnumerable` and 
+`IEnumerator`. The generic versions are preferred for modern code.
+
+# Enumeration sources with iterator methods
```

Comment:
IMHO, I'm not sure that an advanced scenario is appropriate here - the tutorials should be an introduction to the basics of C#, not a comprehensive coverage.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
From: Phillip Carter <notifications@github.com>
Sent: Tuesday, December 1, 2015 2:00:36 PM
To: dotnet/core-docs
Cc: Steve Hoag
Subject: Re: [core-docs] WIP (Work in Progress): Iterator tutorials (#172)


In docs/languages/csharp-iterators.md<https://github.com/dotnet/core-docs/pull/172#discussion_r46345915>:

> +foreach (var item in collection)
> +{
> +   Console.WriteLine(item.ToString());
> +}
> +```
> +
> +That's all there is to it. To iterate over all the contents of a collection,
> +the `foreach` statement is all you need. The `foreach` statement isn't magic,
> +though. It relies on two interfaces defined in the .NET core library in order
> +to generate the code necessary to iterate a collection: `IEnumerable<T>` and
> +`IEnumerator<T>`.
> +
> +Both of these interfaces also have non-generic counterparts: `IEnumerable` and
> +`IEnumerator`. The generic versions are preferred for modern code.
> +
> +# Enumeration sources with iterator methods


I mentioned this above, but I think a slightly more advanced scenario could go well here. Here's a very rough scenario (feel free to suggest a better one):

For example, say I had a collection of objects which I wanted to turn into a CSV that a user could download on-demand.

A first approach would serialize the list as a CSV and then write that CSV to the response of a web controller. This would work well with a small collection of items, but what if I had a list of a few thousand objects? What of these objects had large values in them as well? Memory would get slammed by this! (fun fact: I've run into an OutOfMemoryException with a similar scenario in the past)

Iterators to the rescue! Using yield return to defer execution allows for an object --> CSV row pipeline that each item in the collection runs through. This "streaming" of objects through the CSV-to-response pipeline will allow your code to scale without consuming too much memory when your collection of objects grows.

-
Reply to this email directly or view it on GitHub<https://github.com/dotnet/core-docs/pull/172/files#r46345915>.